model/histogram/generic.go

```diff
@@ -540,3 +540,86 @@ var exponentialBounds = [][]float64{
 		0.9892280131939752, 0.9919100824251095, 0.9945994234836328, 0.9972960560854698,
 	},
 }
+
+func shrink[IBC InternalBucketCount](originSpans []Span, originBuckets []IBC, originSchema, targetSchema int32, deltaBuckets bool) ([]Span, []IBC) {
+	var (
+		targetSpans           []Span // The spans in the target schema.
+		targetBuckets         []IBC  // The buckets in the target schema.
+		bucketIdx             int32  // The index of bucket in the origin schema.
+		lastBucketCount       IBC    // The last visited bucket's count in the origin schema.
+		lastTargetBucketIdx   int32  // The index of the last added target bucket.
+		lastTargetBucketCount IBC
+		origBucketIdx         int // The position of a bucket in originBuckets slice.
+	)
+
+	for _, span := range originSpans {
+		// Determine the index of the first bucket in this span.
+		bucketIdx += span.Offset
+		for j := 0; j < int(span.Length); j++ {
+			// Determine the index of the bucket in the target schema from the index in the original schema.
+			targetBucketIdx := targetIdx(bucketIdx, originSchema, targetSchema)
+
+			switch {
+			case len(targetSpans) == 0:
+				// This is the first span in the targetSpans.
+				span := Span{
+					Offset: targetBucketIdx,
+					Length: 1,
+				}
+				targetSpans = append(targetSpans, span)
+				targetBuckets = append(targetBuckets, originBuckets[0])
+				lastTargetBucketIdx = targetBucketIdx
+				lastBucketCount = originBuckets[0]
+				lastTargetBucketCount = originBuckets[0]
+
+			case lastTargetBucketIdx == targetBucketIdx:
+				// The current bucket has to be merged into the same target bucket as the previous bucket.
+				if deltaBuckets {
+					lastBucketCount += originBuckets[origBucketIdx]
+					targetBuckets[len(targetBuckets)-1] += lastBucketCount
+					lastTargetBucketCount += lastBucketCount
+				} else {
+					targetBuckets[len(targetBuckets)-1] += originBuckets[origBucketIdx]
+				}
+
+			case (lastTargetBucketIdx + 1) == targetBucketIdx:
+				// The current bucket has to go into a new target bucket,
+				// and that bucket is next to the previous target bucket,
+				// so we add it to the current target span.
+				targetSpans[len(targetSpans)-1].Length++
+				lastTargetBucketIdx++
+				if deltaBuckets {
+					lastBucketCount += originBuckets[origBucketIdx]
+					// 这里可以这么用吗
+					targetBuckets = append(targetBuckets, lastBucketCount-lastTargetBucketCount)
+					lastTargetBucketCount = lastBucketCount
+				} else {
+					targetBuckets = append(targetBuckets, originBuckets[origBucketIdx])
+				}
+
+			case (lastTargetBucketIdx + 1) < targetBucketIdx:
```

Comment:
Maybe not that important , @beorn7  will tell. I think there's one more case. Assuming to consecutive spans are 3 or 4 buckets apart, reducing the schema may shrink the gap to 1 or 2, meaning empty buckets should be introduced. 

See https://github.com/prometheus/client_golang/blob/27f0506d6ebbb117b6b697d0552ee5be2502c5f2/prometheus/histogram.go#L1296
Also my PR in opentelemetry: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24026/files

Reply:
See the thoughts formulated [here](https://github.com/prometheus/prometheus/blob/88de3d4491f971fca41a8fe018d52d6ddd204647/model/histogram/float_histogram.go#L376-L392). 

In short, I think that we only should use empty buckets to merge spans in the exposition format, but not internal in Prometheus and in the TSDB.

The `maxEmptyBuckets` argument is still in the code, but it is set to zero everywhere in non-test code.

So we don't need to treat this case specially here, at least for now. Should we ever feel there is a need, we can still introduce it later.

Reply:
Ok, that makes sense, also it actually helps in my use case since empty buckets might keep a histogram above the bucket limit without a reason.

Reply:
To make sure I understand this correctly, we should keep the implementation for now, right? Which means we do not  introduce empty buckets/spans when there exists gap.
@krajorama @beorn7